From 9c1ffa5e821d8c49f3114193fb30bdfd910c584a Mon Sep 17 00:00:00 2001
From: Jagpal Singh Gill <paligill@gmail.com>
Date: Wed, 17 Sep 2025 15:12:11 -0700
Subject: [PATCH 13/13] leakdetector: check event before event_read()

The event_read() does not clear the EPOLLIN flag immediately; it is
cleared during the subsequent epoll check. This may cause issue for a
pending EPOLLIN clear, where the event callback is called but no event
to read. Hence, call event_wait with a zero timeout to ensure that
event_read() is only invoked when an event is available, preventing it
from blocking. This issue may show in scale tests where events are
triggered too often.

Tested:
Tested on Meta's Ventura platform by changing GPIO leak detector state
frequently.

Change-Id: I6a43e469c124d0de8aa56174a1f3d898ba35c825
Signed-off-by: Jagpal Singh Gill <paligill@gmail.com>
---
 src/GPIOInterface.cpp | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/GPIOInterface.cpp b/src/GPIOInterface.cpp
index 0bd0f4c..81a93d3 100644
--- a/src/GPIOInterface.cpp
+++ b/src/GPIOInterface.cpp
@@ -4,6 +4,7 @@
 #include <phosphor-logging/lg2.hpp>
 #include <sdbusplus/async.hpp>
 
+#include <chrono>
 #include <exception>
 #include <memory>
 #include <stdexcept>
@@ -81,6 +82,15 @@ auto GPIOInterface::readGPIOAsyncEvent() -> sdbusplus::async::task<>
         // Wait for the fd event for the line to change
         co_await fdioInstance->next();
 
+        // event_read() does not clear the EPOLLIN flag immediately; it is
+        // cleared during the subsequent epoll check. Therefore, call event_wait
+        // with a zero timeout to ensure that event_read() is only invoked when
+        // an event is available, preventing it from blocking.
+        if (!line.event_wait(std::chrono::milliseconds(0)))
+        {
+            continue;
+        }
+
         line.event_read();
         auto lineValue = line.get_value();
 
-- 
2.47.3

